### PR TITLE
Add support for create, update, and delete for custom role definitions

### LINKED ISSUE #16487: Add support for custom role definitions
Once Azure/azure-rest-api-specs#11465 is checked in, regenerate source to support custom role definitions.

### PR DESCRIPTION
**Pull request by christothes**
closes #16487

### CODE REVIEW

File: eng/common/TestResources/README.md [hunk 1]
```diff
@@ -30,7 +30,7 @@ Connect-AzAccount -Subscription 'YOUR SUBSCRIPTION ID'
 eng\common\TestResources\New-TestResources.ps1 -ServiceDirectory 'search'
 ```
 
-The `OutFile` switch would be set if you are running this for a .NET project on Windows. This will save test environment settings
+The `OutFile` switch should be set if you are running this for a .NET project on Windows. This will save test environment settings
```
Comment (weshaggard) [inline eng/common/TestResources/README.md:33]: Changes under "eng/common" should be made in the azure-sdk-tools repo. https://github.com/Azure/azure-sdk-tools/blob/master/doc/common/common_engsys.md#workflow

Comment (christothes) [reply eng/common/TestResources/README.md:33]: Thanks, Wes. I'll move this change over there.

File: sdk/keyvault/Azure.Security.KeyVault.Administration/tests/AccessControlClientLiveTests.cs [hunk 2]
```diff
@@ -32,6 +32,53 @@ public async Task GetRoleDefinitions()
...
+        {
+            var description = Recording.GenerateAlphaNumericId("role");
+            var name = Recording.Random.NewGuid();
+            var originalPermissions = new KeyVaultPermission();
+            originalPermissions.DataActions.Add(KeyVaultDataAction.BackupHsmKeys);
+
+            KeyVaultRoleDefinition createdDefinition = await Client.UpsertRoleDefinitionAsync(description, originalPermissions, KeyVaultRoleScope.Global, name);
+            await Client.DeleteRoleDefinitionAsync(name, KeyVaultRoleScope.Global);
+
+            List<KeyVaultRoleDefinition> results = await Client.GetRoleDefinitionsAsync(KeyVaultRoleScope.Global).ToEnumerableAsync().ConfigureAwait(false);
```
Comment (maorleger) [inline sdk/keyvault/Azure.Security.KeyVault.Administration/tests/AccessControlClientLiveTests.cs:77]: I noticed in passing that there's no `Client.GetRoleDefinition{Async}(scope, name)` and was wondering if that's something we want to purposely omit? I'm implementing it in JS so I figured it's worth asking (maybe I missed it). Thanks! 

https://github.com/Azure/azure-rest-api-specs/blob/master/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.2-preview/rbac.json#L132

Comment (christothes) [reply sdk/keyvault/Azure.Security.KeyVault.Administration/tests/AccessControlClientLiveTests.cs:77]: Good catch  - I missed this operation!

File: sdk/keyvault/Azure.Security.KeyVault.Administration/api/Azure.Security.KeyVault.Administration.netstandard2.0.cs [hunk 3]
```diff
@@ -28,16 +28,20 @@ public partial class KeyVaultAccessControlClient
...
+        public virtual Azure.Pageable<Azure.Security.KeyVault.Administration.KeyVaultRoleDefinition> GetRoleDefinitions(Azure.Security.KeyVault.Administration.Models.KeyVaultRoleScope roleScope, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public virtual Azure.AsyncPageable<Azure.Security.KeyVault.Administration.KeyVaultRoleDefinition> GetRoleDefinitionsAsync(Azure.Security.KeyVault.Administration.Models.KeyVaultRoleScope roleScope, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public virtual Azure.Response<Azure.Security.KeyVault.Administration.KeyVaultRoleDefinition> UpsertRoleDefinition(string roleDefinitionDescription, Azure.Security.KeyVault.Administration.KeyVaultPermission permissions, Azure.Security.KeyVault.Administration.Models.KeyVaultRoleScope roleScope = default(Azure.Security.KeyVault.Administration.Models.KeyVaultRoleScope), System.Guid? roleDefinitionName = default(System.Guid?), System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Administration/api/Azure.Security.KeyVault.Administration.netstandard2.0.cs:43]: Is "Upsert" okay to use? In search, we use "CreateOrUpdate" as a prefix. @tg-msft @schaabs @KrzysztofCwalina what do you think? I don't see either in https://azure.github.io/azure-sdk/dotnet_introduction.html#sync-and-async.

File: sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAccessControlClient.cs [hunk 4]
```diff
@@ -151,11 +151,138 @@ public virtual AsyncPageable<KeyVaultRoleDefinition> GetRoleDefinitionsAsync(Key
...
+                    Description = roleDefinitionDescription,
+                    RoleName = name,
+                    RoleType = KeyVaultRoleType.CustomRole
+                };
+                properties.AssignableScopes.Add(roleScope);
+                properties.Permissions.Add(permissions);
+
+                var parameters = new RoleDefinitionCreateParameters(properties);
+
+                return await _definitionsRestClient.CreateOrUpdateAsync(
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAccessControlClient.cs:181]: Given the generator is using `CreateOrUpdate`, I'd throw my $0.02 USD behind that name.

File: sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultRoleScope.cs [hunk 5]
```diff
@@ -1,31 +1,14 @@
...
 
-namespace Azure.Security.KeyVault.Administration
+namespace Azure.Security.KeyVault.Administration.Models
 {
-    /// <summary>
-    /// A scope of the role assignment.
-    /// </summary>
-    public readonly struct KeyVaultRoleScope : IEquatable<KeyVaultRoleScope>
+    [CodeGenModel("RoleScope")]
+    public readonly partial struct KeyVaultRoleScope : IEquatable<KeyVaultRoleScope>
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultRoleScope.cs:10]: Nit: FWIW you don't need to "re-implement" any interfaces on a partial. The generated one will take care of that.

File: sdk/keyvault/Azure.Security.KeyVault.Administration/src/RoleDefinitionType.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Azure.Core;
+
+namespace Azure.Security.KeyVault.Administration.Models
+{
+    [CodeGenModel("RoleDefinitionType")]
+    public readonly partial struct KeyVaultRoleDefinitionType : IEquatable<KeyVaultRoleDefinitionType>
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Administration/src/RoleDefinitionType.cs:10]: Nit: same here and elsewhere - you don't have to "re-implement" interfaces or anything. I'd recommend not so the generator does whatever it needs to uniformly.

**Review by heaths — APPROVED** [review]

File: sdk/keyvault/Azure.Security.KeyVault.Administration/api/Azure.Security.KeyVault.Administration.netstandard2.0.cs (hunk 3 above)
Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Administration/api/Azure.Security.KeyVault.Administration.netstandard2.0.cs:43]: @maorleger asked about this at https://github.com/Azure/azure-sdk-for-js/issues/12206#issuecomment-769482971 and I got to thinking there: `CreateOrUpdate` is quite obvious in its intent, where some users may not know what `Upsert` means. I still defer to @KrzysztofCwalina (@tg-msft did offline as well) but we should get it recorded in our guidelines. Seems [TypeScript allows](https://azure.github.io/azure-sdk/typescript_design.html#ts-client-naming-conventions) `upsert` as well as `set`, which is another option to consider here. I think that one is less obvious than `CreateOrUpdate`, though.

File: eng/common/TestResources/README.md (hunk 1 above)
Comment (christothes) [reply eng/common/TestResources/README.md:33]: You’re right. I had forgotten that we default now.

File: sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAccessControlClient.cs (hunk 4 above)
Comment (christothes) [reply sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAccessControlClient.cs:181]: We used Upsert in tables and it sounds more succinct to me. Happy to go with the majority / guidelines though.

Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAccessControlClient.cs:181]: `CreateOrUpdate` is far more prevalent in track 2 sources in this repo:

```powershell
$matches = dir sdk\* | dir -include Azure.* -exclude Azure.ResourceManager.*, Azure.Management.* -dir `
  | join-path -child src -resolve | dir -filter *.cs -recurse | select-string 'Upsert|CreateOrUpdate' `
  | group @{e={$_.Matches[0].Value}} -noelem
```
```text
Count Name
----- ----
  215 CreateOrUpdate
   42 upsert
```

Comment (tg-msft) [reply sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAccessControlClient.cs:227]: `Upsert` tends to mean different things to different folks, unfortunately.  At this point I learn toward `CreateOrReplace`/`CreateOrUpdate` and `CreateOrMerge` as two separate names.

Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAccessControlClient.cs:227]: I have an internal thread on this and sounds like we're converging toward:

* Set: creates or replaces a resource entirely.
* Upsert: creates or updates a resources in a database.
* CreateOrUpdate: creates or updates a resource in something other than a database.

Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAccessControlClient.cs:227]: "Set" probably isn't good in most cases. It seems to work for AppConfig values or KV secrets where the whole things is replaced (ignoring metadata), but seldom else.

**Pull request merged**